cores/arduino/stm32/analog.c

```diff
@@ -110,10 +112,12 @@
   * @}
   */
 
-/** @addtogroup STM32F4xx_System_Private_Variables
-  * @{
-  */
+ADC_HandleTypeDef AdcHandleTab[ADC_NUM];
+static DMA_HandleTypeDef DmaHandleTab[ADC_NUM] = {NULL};
+static PinName g_current_pinTab[ADC_NUM];
 static PinName g_current_pin = NC;
+void (*cb[ADC_NUM])(void *user_data);
+static bool use_dma = false;
 
```

Comment:
Maybe it would be useful to have a global ADC structure starting which would contain the ADC handle, the DMA handle,  cb and also use_dma I guess. 

+ADC_HandleTypeDef AdcHandleTab[ADC_NUM];
 +static DMA_HandleTypeDef DmaHandleTab[ADC_NUM] = {NULL};
 +static PinName g_current_pinTab[ADC_NUM];
 +void (*cb[ADC_NUM])(void *user_data);
 +static bool use_dma = false;

If you define your structure in such way that ADC_HandleTypeDef is the first structure parameter, then you can also access the global structure based on the handle. 

something like 
typedef struct {
  ADC_HandleTypeDef AdcHandle;
  DMA_HandleTypeDef DmaHandle;
  PinName current_pin;
  void(*cb)(void *user_data);
  void* user_data;
  bool use_dma;
} adc_context;

static adc_context context(ADC_NUM];

then in MSp_Init/Deinit and also in HAL_ADC_ConvCpltCallback you can get all information like cb: 

void HAL_ADC_ConvCpltCallback(ADC_HandleTypeDef *AdcHandle) {
  adc_context* context = (adc_context *) AdcHandle;
  if(context->cb != NULL)
    context->cb(context->user_data);
}

Reply:
Done. New release available.